### PR TITLE
CDXML format support

### PR DESCRIPTION
**Pull request by even1024**
All basic features and fragments' connections supported.

### CODE REVIEW

File: utils/indigo-depict/main.c [hunk 1]
```diff
@@ -899,6 +902,7 @@ int main(int argc, char* argv[])
         }
         else
         {
+            indigoLayout(obj);
```
Comment (mkviatkovskii) [inline utils/indigo-depict/main.c:905]: Why do we need to layout in this case? We layout in other branch of condition.

File: utils/indigo-depict/main.c [hunk 2]
```diff
@@ -384,6 +386,7 @@ int parseParams(Params* p, int argc, char* argv[])
     indigoSetOptionBool("treat-x-as-pseudoatom", 1);
     indigoSetOptionBool("render-coloring", 1);
     indigoSetOptionBool("render-highlight-color-enabled", 1);
+    indigoSetOption("molfile-saving-mode", "3000");
```
Comment (mkviatkovskii) [inline utils/indigo-depict/main.c:389]: This change also looks pretty strange, do we really need it?

**Review by mkviatkovskii — CHANGES_REQUESTED** [review]
We also need to add tests for CDXML loader, don't see any.

**Review by mkviatkovskii — APPROVED** [review]

File: api/tests/integration/tests/formats/cdxml_to_mol.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+files.sort()
+for filename in files:
+    print(filename)
+    try:
+        mol = indigo.loadMoleculeFromFile(os.path.join(root, filename))
+        print(mol.molfile())
+    except IndigoException as e:
+        print(getIndigoExceptionText(e))
+        print("*** Try as Query ***")
+        mol = indigo.loadQueryMoleculeFromFile(os.path.join(root, filename))
```
Comment (SunFellow) [inline api/tests/integration/tests/formats/cdxml_to_mol.py:22]: Why not to catch and print exception here as well? Even caught the exception can fail the test.
In addition, maybe it makes sense to complete the cycle by files to get whole diagnostics?

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        {
+            return std::to_string( val );
+        }
+
+    private:
+        int val;
+    };
+
+
+    struct _ExtConnection
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:77]: Please check the naming style: underscore is usually used for inner protected types. But here `struct _ExtConnection` is independent (not a subclass), exposed in the header, thus it's quite public.

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    struct CdxmlNode
+    {
+        CdxmlNode() : element(ELEM_C), type( kCDXNodeType_Element ) // Carbon by default
+        {
+        }
+        AutoInt id;
+        std::string label;
+        AutoInt element;
+        Vec3f pos;
+        int type;
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:93]: is not initialized by default. (is not assigned in constructor and does not have own default constructor)

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        AutoInt isotope;
+        AutoInt charge;
+        AutoInt radical;
+        AutoInt valence;
+        AutoInt hydrogens;
+        AutoInt stereo;
+        AutoInt enchanced_stereo;
+        AutoInt stereo_group;
+        AutoInt index;
+        bool is_not_list;
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:103]: is not initialized by default. (is not assigned in constructor and does not have own default constructor)

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    {
+        CdxmlBond() : order(1)
+        {
+        }
+        AutoInt id;
+        std::pair<AutoInt, AutoInt> be;
+        AutoInt order;
+        AutoInt stereo;
+        AutoInt dir;
+        bool swap_bond;
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:120]: is not initialized by default. (is not assigned in constructor and does not have own default constructor)

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    };
+
+    struct CdxmlBracket
+    {
+        CdxmlBracket() : repeat_pattern(RepeatingUnit::HEAD_TO_TAIL)
+        {
+        }
+        std::vector<AutoInt> bracketed_list;
+        int usage;
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:129]: is not initialized by default. (is not assigned in constructor and does not have own default constructor)

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        std::vector<std::string> strings;
+        size_t start;
+        size_t end = 0;
+        while ((start = str.find_first_not_of(delim, end)) != std::string::npos)
+        {
+            end = str.find(delim, start);
+            strings.push_back(str.substr(start, end - start));
+        }
+        return strings;
+    }
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:147]: IMHO, this is auxiliary function and should not be exposed in the header.

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+        Molecule* _pmol;
+        QueryMolecule* _pqmol;
+        std::unordered_map<int, int> _id_to_atom_idx;
+        std::unordered_map<int, int> _id_to_node_index;
+        std::unordered_map<int, int> _id_to_bond_index;
+        std::vector<int> _fragment_nodes;
+
+    private:
+        MoleculeCdxmlLoader(const MoleculeCdxmlLoader&); // no implicit copy
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:199]: `= delete;`

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+        void _parseCDXMLPage(TiXmlElement* pElem);
+        void _parseCDXMLFragment(TiXmlElement* pElem);
+        void _parseFragmentAttributes(TiXmlAttribute* pAttr);
+
+        void _appendQueryAtom(const char* atom_label, std::unique_ptr<QueryMolecule::Atom>& atom);
+        void _updateConnection(const CdxmlNode& node, int atom_idx);
+
+        Molecule* _pmol;
+        QueryMolecule* _pqmol;
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:192]: It makes sense to initialize them by default with null_ptr-s

File: core/indigo-core/molecule/src/molecule_cdxml_loader.cpp [hunk 12]
```diff
@@ -0,0 +1,668 @@
...
+{
+    auto cdxml_bbox_lambda = [this](std::string& data) {
+        std::vector<std::string> coords = split(data, ' ');
+        if (coords.size() == 4)
+        {
+            this->_has_bounding_box = true;
+            this->_cdxml_bbox = Rect2f(Vec2f(std::stof(coords[0]), std::stof(coords[1])), Vec2f(std::stof(coords[2]), std::stof(coords[3])));
+        }
+        else
+            throw Error("Not enought coordinates for atom position");
```
Comment (SunFellow) [inline core/indigo-core/molecule/src/molecule_cdxml_loader.cpp:115]: Mistype: `enought`

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    public:
+        DECL_ERROR;
+
+        MoleculeCdxmlLoader(Scanner& scanner);
+
+        void loadMolecule(BaseMolecule& mol);
+
+        StereocentersOptions stereochemistry_options;
+        bool ignore_bad_valence;
+        bool _has_bounding_box;
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:160]: is not initialized by default. (is not assigned in constructor and does not have own default constructor)

File: core/indigo-core/molecule/src/molecule_cdxml_loader.cpp [hunk 14]
```diff
@@ -0,0 +1,668 @@
...
+    _pmol = NULL;
+    _pqmol = NULL;
+    if (mol.isQueryMolecule())
+    {
+        _pqmol = &mol.asQueryMolecule();
+    }
+    else
+    {
+        _pmol = &mol.asMolecule();
+        _pmol->setIgnoreBadValenceFlag(ignore_bad_valence);
```
Comment (SunFellow) [inline core/indigo-core/molecule/src/molecule_cdxml_loader.cpp:58]: Passing a variable that was never initialized or assigned.

File: core/indigo-core/molecule/molecule_cdxml_loader.h [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+        void _parseCDXMLAttributes(TiXmlAttribute* pAttr);
+        void _parseNode(CdxmlNode& node, TiXmlElement* pElem);
+        void _addNode(CdxmlNode& node);
+
+        void _parseBond(CdxmlBond& bond, TiXmlAttribute* pAttr);
+        void _addBond(CdxmlBond& node);
+
+        void _parseBracket(CdxmlBracket& bracket, TiXmlAttribute* pAttr);
+
+        void _applyDispatcher(TiXmlAttribute* pAttr, const std::unordered_map<std::string, std::function<void(std::string&)>>& dispatcher);
```
Comment (SunFellow) [inline core/indigo-core/molecule/molecule_cdxml_loader.h:179]: Such a wordy type (`std::unordered_map<std::string, std::function<void(std::string&)>>& dispatcher`) is worth using or typedef.

File: core/indigo-core/molecule/molecule_cdxml_loader.h (hunk 8 above)
Comment (SunFellow) [reply core/indigo-core/molecule/molecule_cdxml_loader.h:129]: 'usage' essentially has `enum CDXBracketUsage` type. Better utilize C++ strong typing.

**Pull request merged**